docs/builtins.md

```diff
@@ -0,0 +1,806 @@
+
+
+If the width/precision is preceded by a `-`, the value is padded with
+spaces on the right rather than the left. If it is preceded by a 0, the
+value is padded with leading zeros rather than spaces.
+
+A specifier width/precision and alignment prefix (`-` or `0`) can be
+used with all valid specifiers. For example:
+
+```evy
+printf "right:  |%7.2f|\n" 1
+printf "left:   |%-7.2v|\n" "abcd"
+printf "zeropad:|%07.2f|\n" 1.2345
+```
+Output
+```
+right:  |   1.00|
+left:   |ab     |
+zeropad:|0001.23|
+```
+
+## Types
+
+### `len`
+
+`len` returns the number of characters in a string, the number of
+elements in an array or the number of key-value pairs in a map.
+
+#### Example
+
+```evy
+l := len "abcd"
+print "len \"abcd\":" l
+
+l = len [1 2]
+print "len [1 2]:" l
+
+l = len {a:3 b:4 c:5}
+print "len {a:3 b:4 c:5}:" l
+```
+
+Output
+```
+len "abcd": 4
+len [1 2]: 2
+len {a:3 b:4 c:5}: 3
+```
+
+#### Reference 
+
+    len:num a:any
+
+The `len` function takes a single argument, which can be a string, an
+array, or a map. If the argument is a string, `len` returns the number
+of characters in the string. If the argument is an array, `len` returns
+the number of elements in the array. If the argument is a map, `len`
+returns the number of key-value pairs in the map. If the argument is of
+any other type, `len` will cause a runtime error.
+
+### `typeof`
+
+`typeof` returns the type of the argument as string value.
+
+#### Example
+
+```evy
+a:any
+a = "abcd"
+t := typeof a
+print "typeof \"abcd\":" t
+
+t = typeof {kind:true strong:true}
+print "typeof {kind:true strong:true}:" t
+
+t = typeof [[1 2] [3 4]]
+print "typeof [[1 2] [3 4]]:" t
+
+t = typeof [1 2 true]
+print "typeof [1 2 true]:" t
+
+print "typeof []:" (typeof [])
+```
+
+Output
+```
+typeof "abcd": string
+typeof {a:3 b:4 c:5}: {}bool
+typeof [1 2 true]: [][]num
+typeof [1 2 true]: []any
+typeof []: []
+```
+
+#### Reference 
+
+    typeof:string a:any
+
+The `typeof` function takes a single argument, which can be of any type.
+The function returns a string that represents the type of the argument.
+The string returned by `typeof` is the same as the type in an Evy
+program, for example `num`, `bool`, `string`, `[]num`, `{}[]any`. For
+an empty composite literal, `typeof` returns `[]` or `{}` as it can be
+matched to any subtype, e.g. `[]` can be passed to a function that
+takes an argument of `[]num`, or `[]string`.
+
+## Map
+
+### `has`
+
+`has` returns whether given map has a key or not.
+
+#### Example
+
+```evy
+map := {a:1}
+printf "has %v %q: %t\n" map "a" (has map "a")
+printf "has %v %q: %t\n" map "X" (has map "X")
+```
+
+Output
+```
+has {a:1} "a": true
+has {a:1} "X": false
+```
+
+#### Reference 
+
+    has:bool map:{} key:string
+
+The `has` function takes two arguments: a map and a key. The function
+returns true if the map has the key, and false if the map does not have
+the key. The map can have any value type, such as `{}num` or `{}
```

Comment:
nit: I would not split the type across lines. it should still fit and be shorter than the previous line if joined.

Reply:
done.